File: packages/es-markdown-core/src/parse-options.ts
/**
 * The feature switches for one parse, fixed for the document's whole series.
 *
 * Every switch is optional and an omitted one is ON, so the only reason to
 * name one is to turn it off. A switch that is present must be a boolean —
 * including one written as `undefined`, which is a `TypeError` and not a
 * fallback to the default.
 */
export interface ParseOptions {
    /** Replaces straight quotes, dashes, and ellipses with typographic
     * forms. */
    readonly smartPunctuation?: boolean;
    /** Parses footnote definitions and references. */
    readonly footnotes?: boolean;
    /** Parses pipe tables. */
    readonly tables?: boolean;
    /** Parses `~struck~` and `~~struck~~`; the closer must match the opener's
     * tilde count. */
    readonly strikethrough?: boolean;
    /** Recognizes bare URLs and email addresses as links.
     *
     * `<https://x>` is CommonMark's own autolink and is not gated here. */
    readonly autolinks?: boolean;
    /** Parses `[ ]` and `[x]` task-list item markers. */
    readonly taskLists?: boolean;
    /** Parses formula spans and blocks, including dollar and LaTeX delimiters
     * and `formula` fenced blocks. */
    readonly formulas?: boolean;
    /** Parses the directive grammar: `:name` inline, `::name` leaf and
     * `:::name` container at the head of a line, each with its `[label]` and
     * `{attributes}`. */
    readonly directives?: boolean;
    /** Parses cross-links written as `[[reference]]`. */
    readonly crossLinks?: boolean;
    /** Parses embeds written as `![[reference]]`. */
    readonly embeds?: boolean;
}

File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/model/ParseOptions.kt
@file:kotlin.jvm.JvmName("MarkdownCoreKt")
@file:kotlin.jvm.JvmMultifileClass

package com.nouprax.markdown.core

import kotlin.jvm.JvmOverloads
import kotlin.jvm.JvmSynthetic

/**
 * What the parser recognizes.
 *
 * Fixed for a document's whole series: an append carries them forward
 * unchanged, and parsing under different options is a new [Document] rather
 * than a mutation.
 */
public data class ParseOptions
    @JvmOverloads
    constructor(
        /** Replaces straight quotes, dashes, and ellipses with typographic
         * forms. */
        public val smartPunctuation: Boolean = true,
        /** Parses footnote definitions and the references to them. */
        public val footnotes: Boolean = true,
        /** Parses pipe tables. */
        public val tables: Boolean = true,
        /** Parses `~struck~` and `~~struck~~`; the closer must match the
         * opener's tilde count. */
        public val strikethrough: Boolean = true,
        /** Recognizes bare URLs and email addresses as links. */
        public val autolinks: Boolean = true,
        /** Parses `[ ]` and `[x]` task-list item markers. */
        public val taskLists: Boolean = true,
        /** Parses formula spans and blocks: dollar and LaTeX delimiters, and
         * `formula` fenced blocks. */
        public val formulas: Boolean = true,
        /** Parses directives — inline `:name`, and the `::name` and `:::name`
         * block forms. */
        public val directives: Boolean = true,
        /** Parses cross-links written as `[[reference]]`. */
        public val crossLinks: Boolean = true,
        /** Parses embeds written as `![[reference]]`. */
        public val embeds: Boolean = true,
    )

@JvmSynthetic
internal fun ParseOptions.toNativeMask(): Int =
    listOf(
        smartPunctuation,
        footnotes,
        tables,
        strikethrough,
        autolinks,
        taskLists,
        formulas,
        directives,
        crossLinks,
        embeds,
    ).foldIndexed(0) { index, mask, enabled -> if (enabled) mask or (1 shl index) else mask }
